src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+    }
+
+    // TODO: Set additional flags:
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kChangesOmitted)
+}
+
+InteractionModel::AttributeEntry AttributeEntryFrom(const ConcreteClusterPath & clusterPath,
+                                                    const EmberAfAttributeMetadata & attribute)
+{
+    InteractionModel::AttributeEntry entry;
+
+    entry.path = ConcreteAttributePath(clusterPath.mEndpointId, clusterPath.mClusterId, attribute.attributeId);
+    LoadAttributeInfo(entry.path, attribute, &entry.info);
+
+    return entry;
+}
+
+InteractionModel::AttributeEntry AttributeEntryForGlobalListAttribute(const ConcreteClusterPath & clusterPath,
+                                                                      chip::AttributeId globalAttributeId)
+{
+    InteractionModel::AttributeEntry entry;
+
+    entry.path = ConcreteAttributePath(clusterPath.mEndpointId, clusterPath.mClusterId, globalAttributeId);
+    entry.info.flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+
+    return entry;
+}
+
+} // namespace
+
+CHIP_ERROR CodegenDataModel::ReadAttribute(const InteractionModel::ReadAttributeRequest & request,
+                                           InteractionModel::ReadState & state, AttributeValueEncoder & encoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR CodegenDataModel::WriteAttribute(const InteractionModel::WriteAttributeRequest & request,
+                                            AttributeValueDecoder & decoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR CodegenDataModel::Invoke(const InteractionModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                                    InteractionModel::InvokeReply & reply)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+EndpointId CodegenDataModel::FirstEndpoint()
+{
+    // find the first enabled index
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+    for (uint16_t endpoint_idx = 0; endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            return emberAfEndpointFromIndex(endpoint_idx);
+        }
+    }
+
+    // No enabled endpoint found. Give up
+    return kInvalidEndpointId;
+}
+
+std::optional<unsigned> CodegenDataModel::TryFindEndpointIndex(chip::EndpointId id) const
+{
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+
+    if ((mEndpointIterationHint < lastEndpointIndex) && emberAfEndpointIndexIsEnabled(mEndpointIterationHint) &&
+        (id == emberAfEndpointFromIndex(mEndpointIterationHint)))
+    {
+        return std::make_optional(mEndpointIterationHint);
+    }
+
+    // Linear search, this may be slow
```

Comment:
Is there a reason to not just use `emberAfIndexFromEndpoint()` here, instead of reinventing it?